Source/WebKit/UIProcess/ios/PointerTouchCompatibilitySimulator.mm

```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright (C) 2024 Apple Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#import "config.h"
+#import "PointerTouchCompatibilitySimulator.h"
+
+#if PLATFORM(IOS_FAMILY)
+
+#import "UIKitSPI.h"
+#import "UIKitUtilities.h"
+#import "WKBaseScrollView.h"
+#import "WKWebViewInternal.h"
+#import "_WKTouchEventGeneratorInternal.h"
+
+namespace WebKit {
+
+PointerTouchCompatibilitySimulator::PointerTouchCompatibilitySimulator(WKWebView *view)
+    : m_view { view }
+    , m_stateResetWatchdogTimer { RunLoop::main(), this, &PointerTouchCompatibilitySimulator::resetState }
+{
+}
+
+#if HAVE(UISCROLLVIEW_ASYNCHRONOUS_SCROLL_EVENT_HANDLING)
+
+static constexpr auto predominantAxisDeltaRatio = 10;
+static constexpr auto minimumPredominantAxisDelta = 20;
+static constexpr auto stateResetTimerDelay = 50_ms;
+
+static bool hasPredominantHorizontalAxis(CGPoint delta)
+{
+    auto absoluteDeltaX = std::abs(delta.x);
+    return absoluteDeltaX >= minimumPredominantAxisDelta && absoluteDeltaX > std::abs(delta.y * predominantAxisDeltaRatio);
+}
+
+void PointerTouchCompatibilitySimulator::handleScrollUpdate(WKBaseScrollView *scrollView, WKBEScrollViewScrollUpdate *update)
+{
+    RetainPtr view = m_view.get();
+    RetainPtr window = [view window];
+    if (!window) {
+        resetState();
+        return;
+    }
+
+    switch (update.phase) {
+    case WKBEScrollViewScrollUpdatePhaseEnded:
+    case WKBEScrollViewScrollUpdatePhaseCancelled:
+        resetState();
+        return;
+    default:
+        break;
+    }
+
+    if (scrollView._wk_canScrollHorizontallyWithoutBouncing)
+        return;
+
+    auto translation = [update translationInView:view.get()];
```

Comment:
Looks like `translationInView` is not available on visionOS?

Reply:
Ah, in the `!USE(BROWSERENGINEKIT)` codepath the scroll update is just a `UIScrollEvent` (SPI), which uses `-_adjustedAcceleratedDeltaInView:`. Will change to that.